src/sys/socket/addr.rs

```diff
@@ -746,6 +848,16 @@ impl SockAddr {
             SockAddr::Netlink(NetlinkAddr(ref sa)) => (mem::transmute(sa), mem::size_of::<libc::sockaddr_nl>() as libc::socklen_t),
             #[cfg(any(target_os = "ios", target_os = "macos"))]
             SockAddr::SysControl(SysControlAddr(ref sa)) => (mem::transmute(sa), mem::size_of::<sys_control::sockaddr_ctl>() as libc::socklen_t),
+            #[cfg(any(target_os = "android",
+                      target_os = "dragonfly",
+                      target_os = "freebsd",
+                      target_os = "ios",
+                      target_os = "linux",
+                      target_os = "macos",
+                      target_os = "netbsd",
+                      target_os = "openbsd"))]
+            // `libc::sockaddr` and `libc::sockaddr_dl`/`libc::sockaddr_ll`  has diffences size.
```

Comment:
I think it's still correct to cast the datatype to `sockaddr`, though you'll need to do this differently for the `sockaddr_dl` and `sockaddr_ll` platforms. It looks like there are other `libc` APIs that do this kind of thing, which is why the length is being thrown around. I think a better way to look at this is that it's really returning a void pointer rather than one for an actual datatype.

And since we're having this conversation, can you add doccomments to this function?

I'm also wondering if we should alter the return type here to be `(*const libc::sockaddr, libc::socklen_t)`. I think it's more clear how unsafe this is if we leave it a pointer. @asomers any comment on this as you might understand this stuff better than me.